src/kernel.cpp

```diff
@@ -224,6 +224,31 @@ bool cvk_kernel_argument_values::setup_descriptor_sets() {
         descriptor_writes.push_back(writeDescriptorSet);
     }
 
+    if (m_entry_point->program()->uses_printf()) {
+        if (!create_printf_buffer()) {
```

Comment:
It definitely isn't optimal, I did it this way to closely match the existing implementation for program-scope constant data, since notionally the printf buffer is just another program-scope buffer (although not constant). This did get complicated by the possibility of different kernel instances writing to the same buffer, so it doesn't live in the `cvk_program`, but I'm not 100% sure that what I've done is always going to be correct.

Before I make any changes here, do you have any ideas about the best way to handle the buffer? I suspect it makes the most sense for there to be one per program per queue, although if there are multiple kernels in-flight then should the `cvk_executor_thread` actually own the buffer?

Reply:
There's a spectrum of possible options, with different trade-offs. Here are two options with some of the pros and cons:
1. One buffer per kernel instance, i.e. what this PR currently implements. With this option we definitely don't have to worry about concurrent accesses from multiple kernel instances and only concurrency between the invocations of a single kernel instance has to be managed. However this requires allocating one buffer per kernel instance that uses printf which is rather costly. We could add a pool of printf buffers to amortise the cost of those allocations but then we'd have the problem of managing the recycling of the memory (unless we kept as many buffers as the peak number of kernel instances reached; with 1K kernels in flight and a default buffer size of 1MiB, we're already talking about 1GiB of memory). It is possible to restrict the allocation of buffers to those kernels instances that do use printf which would limit the cost but could still result in undesirable pathological cases (e.g. lots of kernel instances making very little use of printf).
2. One buffer per queue. The allocation troubles essentially vanish but we now have to manage concurrent accesses from multiple kernel instances. clvk only supports in-order queues for now but a single kernel instance can result in multiple Vulkan compute dispatches that do run concurrently. Having said that, I think this could be managed correctly with the same approach currently (see clspv PR) used to manage concurrency between the work-items of a kernel instance provided the atomics use a `Device` scope. Alternatively clvk could serialise compute dispatches that make use of the printf buffer. This would also simplify the management of the completion logic to format and print those strings in clvk (only one buffer has to be considered). Maybe we'd want to increase the default size of the buffer a bit, at least until OpenCL standardises additional control over it like in the cl_arm_printf extension for example. The buffer could even be created lazily the first time a kernel that makes use of printf is enqueued (if we do that I'd quite like to see a bitfield declaring features with runtime impact individual kernels use added to the non-semantic instruction set).

Assuming I haven't made a mistake in the above reasoning, I can't really think of a clear downside to going for 2 right now (though it's Sunday evening :-)). Can you think of other issues with this approach?

Reply:
The second option definitely makes the most sense to me, so I've made some changes based on that.

- The printf buffer is now in the queue
- The printf buffer is lazily created when the first kernel that may use printf is enqueued
- Additional pipeline barriers are used to serialize execution of different region dispatches for kernels that use printf
- Kernels that use printf are considered non-batchable. This prevents the case where multiple kernels from different programs try to use printf in the same batched command group, as they may both write to the printf buffer and interfere with each other. 
- The actually printing happens as soon as the command is finished.

I agree with your idea of adding a bitfield to the non-semantic instruction set to allow things to work on a per-kernel basis. Are there any other flags you think should be added other than a 'uses printf' one? I can add any others to my SPIRV-Headers change in the same PR. There are a couple of TODOs in the latest commits that will get resolved when I make those changes.